File: 29Items-Projects/7Algorithms/8-COMPUTATIONAL-GEOMETRY/src/java/src/main/java/org/computationalgeometry/algorithms/ClosestPair.java
package org.computationalgeometry.algorithms;

import java.util.List;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashSet;
import org.computationalgeometry.model.Point2D;

public final class ClosestPair {
    private ClosestPair() {
    }

    public static final class Result {
        private final Point2D first;
        private final Point2D second;
        private final double distance;

        public Result(Point2D first, Point2D second, double distance) {
            this.first = first;
            this.second = second;
            this.distance = distance;
        }

        public Point2D first() {
            return first;
        }

        public Point2D second() {
            return second;
        }

        public double distance() {
            return distance;
        }
    }

    public static Result find(List<Point2D> points) {
        if (points.size() < 2) {
            throw new IllegalArgumentException("closest pair requires at least two points");
        }

        HashSet<Point2D> seen = new HashSet<>();
        for (Point2D point : points) {
            if (!seen.add(point)) {
                return new Result(point, point, 0.0);
            }
        }

        List<Point2D> byX = new ArrayList<>(points);
        byX.sort(Point2D::compareTo);
        List<Point2D> byY = new ArrayList<>(points);
        byY.sort(Comparator.comparingDouble(Point2D::y).thenComparingDouble(Point2D::x));
        PartialResult partial = closestRecursive(byX, byY);
        return new Result(partial.first, partial.second, Math.sqrt(partial.distanceSquared));
    }

    private static PartialResult closestRecursive(List<Point2D> byX, List<Point2D> byY) {
        if (byX.size() <= 3) {
            return bruteForce(byX);
        }

        int midpoint = byX.size() / 2;
        List<Point2D> leftX = new ArrayList<>(byX.subList(0, midpoint));
        List<Point2D> rightX = new ArrayList<>(byX.subList(midpoint, byX.size()));
        double splitX = byX.get(midpoint).x();
        HashSet<Point2D> leftSet = new HashSet<>(leftX);
        List<Point2D> leftY = new ArrayList<>();
        List<Point2D> rightY = new ArrayList<>();
        for (Point2D point : byY) {
            if (leftSet.contains(point)) {
                leftY.add(point);
            } else {
                rightY.add(point);
            }
        }

        PartialResult left = closestRecursive(leftX, leftY);
        PartialResult right = closestRecursive(rightX, rightY);
        PartialResult best = left.distanceSquared <= right.distanceSquared ? left : right;

        List<Point2D> strip = new ArrayList<>();
        for (Point2D point : byY) {
            double dx = point.x() - splitX;
            if (dx * dx < best.distanceSquared) {
                strip.add(point);
            }
        }

        for (int i = 0; i < strip.size(); i++) {
            for (int j = i + 1; j < strip.size() && j < i + 8; j++) {
                double candidate = strip.get(i).distanceSquared(strip.get(j));
                if (candidate < best.distanceSquared) {
                    best = new PartialResult(strip.get(i), strip.get(j), candidate);
                }
            }
        }

        return best;
    }

    private static PartialResult bruteForce(List<Point2D> points) {
        Point2D bestA = points.get(0);
        Point2D bestB = points.get(1);
        double bestDistanceSquared = bestA.distanceSquared(bestB);
        for (int i = 0; i < points.size(); i++) {
            for (int j = i + 1; j < points.size(); j++) {
                double candidate = points.get(i).distanceSquared(points.get(j));
                if (candidate < bestDistanceSquared) {
                    bestA = points.get(i);
                    bestB = points.get(j);
                    bestDistanceSquared = candidate;
                }
            }
        }
        return new PartialResult(bestA, bestB, bestDistanceSquared);
    }

    private static final class PartialResult {
        private final Point2D first;
        private final Point2D second;
        private final double distanceSquared;

        private PartialResult(Point2D first, Point2D second, double distanceSquared) {
            this.first = first;
            this.second = second;
            this.distanceSquared = distanceSquared;
        }
    }
}

File: 29Items-Projects/7Algorithms/8-COMPUTATIONAL-GEOMETRY/src/python/geometry/algorithms/closest_pair.py
from dataclasses import dataclass
from math import sqrt

from geometry.model.point import Point2D


@dataclass(frozen=True)
class ClosestPairResult:
    first: Point2D
    second: Point2D
    distance: float


def closest_pair(points: list[Point2D]) -> ClosestPairResult:
    if len(points) < 2:
        raise ValueError("closest pair requires at least two points")

    seen: set[Point2D] = set()
    for point in points:
        if point in seen:
            return ClosestPairResult(point, point, 0.0)
        seen.add(point)

    points_by_x = sorted(points)
    points_by_y = sorted(points, key=lambda p: (p.y, p.x))
    first, second, distance_squared = _closest_recursive(points_by_x, points_by_y)
    return ClosestPairResult(first, second, sqrt(distance_squared))


def _closest_recursive(
    points_by_x: list[Point2D],
    points_by_y: list[Point2D],
) -> tuple[Point2D, Point2D, float]:
    if len(points_by_x) <= 3:
        return _brute_force(points_by_x)

    midpoint = len(points_by_x) // 2
    left_x = points_by_x[:midpoint]
    right_x = points_by_x[midpoint:]
    split_x = points_by_x[midpoint].x
    left_set = set(left_x)
    left_y = [point for point in points_by_y if point in left_set]
    right_y = [point for point in points_by_y if point not in left_set]

    left_best = _closest_recursive(left_x, left_y)
    right_best = _closest_recursive(right_x, right_y)
    best = left_best if left_best[2] <= right_best[2] else right_best
    best_distance_squared = best[2]

    strip = [
        point
        for point in points_by_y
        if (point.x - split_x) * (point.x - split_x) < best_distance_squared
    ]
    for i, first in enumerate(strip):
        for second in strip[i + 1 : i + 8]:
            candidate = first.distance_squared(second)
            if candidate < best_distance_squared:
                best = (first, second, candidate)
                best_distance_squared = candidate

    return best


def _brute_force(points: list[Point2D]) -> tuple[Point2D, Point2D, float]:
    best_first = points[0]
    best_second = points[1]
    best_distance_squared = best_first.distance_squared(best_second)
    for i, first in enumerate(points):
        for second in points[i + 1 :]:
            candidate = first.distance_squared(second)
            if candidate < best_distance_squared:
                best_first = first
                best_second = second
                best_distance_squared = candidate
    return best_first, best_second, best_distance_squared
